### ISSUE TITLE
incorrectly identifying syntax error in kube.libsonnet

### ISSUE THREAD

**Issue by Datamance** [databricks/intellij-jsonnet#19]
<img width="721" alt="Screen Shot 2019-12-09 at 7 09 57 PM" src="https://user-images.githubusercontent.com/8699411/70547601-2a9c5700-1b3f-11ea-831c-b6d6fb687b02.png">

It looks like the plugin considers an empty pair of curly braces as a syntax error.

**Comment by qeqar** [comment]
found it too. but it is ok to remove them completely.

**Comment by Datamance** [comment]
Sure, but this is a github submodule dependency that I'm not trying to fork in order to get rid of a syntax warning that shouldn't be there in the first place. There's nothing in jsonnet's syntax that suggests overriding with an empty object is a syntax error, and indeed, `kube.libsonnet` compiles just fine :) the plugin just needs to be fixed.

**Comment by bryant1410** [comment]
Do you manage to get a *.libsonnet working? Cause I can't, that's why I sent #12

**Comment by Datamance** [comment]
yeah - you can see it's coloring the code correctly up until the end of that Namespace definition. So you can't get any *.libsonnet working? What version of the plugin do you have?

**Comment by bryant1410** [comment]
> yeah - you can see it's coloring the code correctly up until the end of that Namespace definition. So you can't get any *.libsonnet working? What version of the plugin do you have?

I'm running 0.9, the one available in the marketplace for PyCharm 2019.3.

Have you guys added support for *.libsonnet?

**Comment by qeqar** [comment]
you can add *.libsonnet manually to jsonnet, then it will use the plugin

**Comment by bryant1410** [comment]
> you can add *.libsonnet manually to jsonnet, then it will use the plugin

Oh, yeah, but I think it'd be good for the plugin to support it.

**Comment by Datamance** [comment]
Anyone monitoring this repo? @lihaoyi-databricks ?

**Comment by lihaoyi-databricks** [comment]
Sure I'm monitoring this repo. It should be a trivial fix, but so far nobody has sent a PR that's in a mergeable state, so it hasn't been fixed

**Comment by Datamance** [comment]
@lihaoyi-databricks it's been years since I wrote anything in Java and I've never written an intellij/pycharm plugin, so I wouldn't know where to start. Whereabouts would the trivial fix take place?

**Comment by lihaoyi-databricks** [comment]
@Datamance probably somewhere here https://github.com/databricks/intellij-jsonnet/blob/master/src/com/jsonnetplugin/JsonnetFileTypeFactory.java#L15

**Comment by bryant1410** [comment]
> @Datamance probably somewhere here https://github.com/databricks/intellij-jsonnet/blob/master/src/com/jsonnetplugin/JsonnetFileTypeFactory.java#L15

Wait, are you talking about the *.libsonnet support or the syntax problem described by the OP? Cause the former is tackled by #12

**Comment by Datamance** [comment]
@lihaoyi-databricks see the comment from @bryant1410 above - my issue has nothing to do with *.libsonnet support. It's a parsing bug

**Comment by lihaoyi-databricks** [comment]
@Datamance got it, sorry I must be mixing up issues. The parsing issue should be fixed by making the `objinside` call here https://github.com/databricks/intellij-jsonnet/blob/master/resources/com/jsonnetplugin/Jsonnet.bnf#L39 optional via `objinside?` , similar to the `obj ::=` declaration below it. Do you want to try making the change locally and verifying that it fixes the problem? If so send a PR and I can publish it to the jetbrains package repo

**Comment by Datamance** [comment]
@lihaoyi-databricks I'd love to, but I'm not really sure how to locally test the plugin. Do I just clone this repo, make the changes, and then do "install plugin from disk" ? Do I need to execute a build step first? Or does PyCharm do all the magic?

**Comment by lihaoyi-databricks** [comment]
@Datamance IIRC you need to install the intellij plugin devkit, import the repo into intellij as an intellij platform plugin, right-click on the `.bnf` and `.flex` files and click menu item to generate code, and then you can hit the "package plugin for deployment" button in the "build" dropdown

It's kind of annoying, but that's the way it was when this plugin was written. I heard there's a new CLI-based build system for plugins but we're not on that yet

**Comment by Datamance** [comment]
Alright, I don't have IntelliJ on this machine - hacking on this would be a fun weekend project but unfortunately I've got a bit too much on my plate to spend the afternoon figuring this out.

Since it's a small change, and you already have the proper set up, do you think you could test it out quickly?

FWIW, if you'd like other people to drive the project by contributing, I'd recommend having a fleshed out "Contributing" section in your readme, and supply your reader with directions and links on how to get set up developing PyCharm plugins.
